File: python/102-Binary Tree Level Order Traversal.py
from typing import List
from bintree_helper import *
from collections import deque

class Solution:
    def levelOrder(self, root) -> List[List[int]]:
        "Level by Level Traversal using BFS, Time: O(n), Space: O(n)"
        if root is None: return []
        q = deque([root])
        result = []
        while q:
            sz = len(q)
            result.append([0]*sz)
            for i in range(sz):
                node = q.popleft()
                result[-1][i] = node.val
                if node.left: q.append(node.left)
                if node.right: q.append(node.right)
        return result
    # def levelOrder(self, root) -> List[List[int]]:
    #     "Level by Level Traversal using DFS, Time: O(n), Space: O(n)"
    #     result = []
    #     def dfs(node, level):
    #         if node is None: return
    #         if len(result) == level: result.append([])
    #         result[level].append(node.val)
    #         if node.left: dfs(node.left, level+1)
    #         if node.right: dfs(node.right, level+1)
    #         pass
    #     dfs(root, 0)
    #     return result


root = deserialize("[3,9,20,null,null,15,7]")
result = Solution().levelOrder(root)
assert result == [
  [3],
  [9,20],
  [15,7]
]

File: cpp/102. Binary Tree Level Order Traversal.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
typedef vector<int> vi;
typedef vector<vector<int>> vvi;
typedef pair<int, int> pi;


struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
// BFS, Time: O(n), Space: O(n)
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        if(!root) return vvi();
        queue<TreeNode*> q; q.push(root);
        vvi ans;
        while(q.size()) {
            int qSize = q.size();
            ans.push_back(vi(qSize));
            vi &last = ans.back();
            for(int i = 0; i < qSize; i++) {
                TreeNode *node = q.front(); q.pop();
                last[i] = node->val;
                if(node->left) q.push(node->left);
                if(node->right) q.push(node->right);
            }
        }
        return ans;
    }
};

int main() {
    ios::sync_with_stdio(0); 
    cin.tie(0);

}